esp-hal-common/src/rtc_cntl.rs

```diff
@@ -1,46 +1,608 @@
+
+    /// Get main XTAL frequency
+    /// This is the value stored in RTC register RTC_XTAL_FREQ_REG by the bootloader, as passed to
+    /// rtc_clk_init function.
+    fn get_xtal_freq() -> XtalClock {
+        let rtc_cntl = unsafe { &*RTC_CNTL::ptr() };
+        let xtal_freq_reg = rtc_cntl.store4.read().bits();
+
+        // Values of RTC_XTAL_FREQ_REG and RTC_APB_FREQ_REG are stored as two copies in
+        // lower and upper 16-bit halves. These are the routines to work with such a
+        // representation.
+        let clk_val_is_valid = |val| {
+            (val & 0xffffu32) == ((val >> 16u32) & 0xffffu32) && val != 0u32 && val != u32::MAX
+        };
+        let reg_val_to_clk_val = |val| val & u16::MAX as u32;
+
+        if !clk_val_is_valid(xtal_freq_reg) {
+            return XtalClock::RtcXtalFreq40M;
+        }
+
+        match reg_val_to_clk_val(xtal_freq_reg) {
+            40 => XtalClock::RtcXtalFreq40M,
+            #[cfg(any(feature = "esp32c3", feature = "esp32s3"))]
+            32 => XtalClock::RtcXtalFreq32M,
+            #[cfg(feature = "esp32")]
+            26 => XtalClock::RtcXtalFreq26M,
+            #[cfg(feature = "esp32")]
+            24 => XtalClock::RtcXtalFreq24M,
+            other => XtalClock::RtcXtalFreqOther(other),
+        }
+    }
+
+    /// Get the RTC_SLOW_CLK source
+    fn get_slow_freq() -> RtcSlowClock {
+        let rtc_cntl = unsafe { &*RTC_CNTL::ptr() };
+        let slow_freq = rtc_cntl.clk_conf.read().ana_clk_rtc_sel().bits();
+        match slow_freq {
+            0 => RtcSlowClock::RtcSlowClockRtc,
+            1 => RtcSlowClock::RtcSlowClock32kXtal,
+            2 => RtcSlowClock::RtcSlowClock8mD256,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Select source for RTC_SLOW_CLK
+    fn set_slow_freq(slow_freq: RtcSlowClock) {
+        unsafe {
+            let rtc_cntl = &*RTC_CNTL::ptr();
+            rtc_cntl.clk_conf.modify(|_, w| {
+                w.ana_clk_rtc_sel()
+                    .bits(slow_freq as u8)
+                    // Why we need to connect this clock to digital?
+                    // Or maybe this clock should be connected to digital when
+                    // XTAL 32k clock is enabled instead?
+                    .dig_xtal32k_en()
+                    .bit(match slow_freq {
+                        RtcSlowClock::RtcSlowClock32kXtal => true,
+                        _ => false,
+                    })
+                    // The clk_8m_d256 will be closed when rtc_state in SLEEP,
+                    // so if the slow_clk is 8md256, clk_8m must be force power on
+                    .ck8m_force_pu()
+                    .bit(match slow_freq {
+                        RtcSlowClock::RtcSlowClock8mD256 => true,
+                        _ => false,
+                    })
+            });
+
+            esp_rom_delay_us(300u32);
+        };
+    }
+
+    /// Select source for RTC_FAST_CLK
+    fn set_fast_freq(fast_freq: RtcFastClock) {
```

Comment:
this gives `warning: associated function is never used:set_fast_freq` when not targeting ESP32-C3 currently

Reply:
Add actual implementation of `rtc::configure_clock()` for the other chips, the warnings are gone now.